### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by schernichkin** [litherum/cereal-conduit#4]
Hi! I’d like to discuss some things regarding cereal-conduit. First of all, I’m taking little different approach:
1.  I’m trying to avoid exceptions wherever it possible (I really hate exceptions, especially in pure functional languages ;)
2.  I’m trying to return NeedInput ASAP (I don’t want to feed to runGetPartial an empty steam before starting receiving any data, I’m _sure_ I need some input to do some sensible job. Moreover, I find runGetPartial little bit counterintuitive: on first call it accept empty stream nicely, but on subsequent calls it use it to indicate end of input)

I’d like to know you opinion on my approaches and probably some comments on your ones. Just to know, if we could cooperate in development. You may also take a look at my code: https://github.com/schernichkin/cereal-conduit. I will not issue pull requests for a while, because our approaches quite different now and I afraid we will inhibit each other.

Second, I’ve noticed you accumulate partially consumed stream in conduitGet to return it as unconsumed in case of early termination. The idea is nice and I like it, but it have some drawbacks as well. First, one of design goals of the Conduits is to prevent space leaks on input. So we should be very careful while accumulating anything. But I think it’s not a case here, because Cereal will accumulate partially parsed data any way. Second, sinkGet currently does not obey this idea; it returns nothing on early termination and last chunk on fail. I may modify it if you’d like. Finally, it will be quite hard for me to create conduitGet which don’t raise exceptions and obeys this rule. The first idea comes in my mind is to return Done and close conduit in case of failure. And it quite reasonable – when we reading stream of messages and find ourselves unable to read another message we should stop reading because it very unlikely we can proceed somehow. But conduit has type is Pipe i o m () , there is no way to stream up information why it was closed, text message from Cereal will be lost. Probably logging could help?

I think discussing practical applications will help us make design decisions better. As for me – I’m intending to use cereal-conduit to read stream of messages from unreliable source (UPD broadcast). My system should pick message and process it, each messages has serial number, if system can’t read message (fairly common case) it should stop processing stream and ask source to retranslate all messages from last successfully parsed one till current. I did not intend to use something like conduitGet, sinkGet had all I needed. And I would be pleased to know what you using cereal-conduit for and what behavior excepting from it. Thanks.

**Comment by litherum** [comment]
Thanks for showing interest in this.

Returning a type of Either Error Result makes a lot of sense for Sinks, but not as much for Conduits, because stream synchronization (meaning: where one item in the stream ends and the next one ends) may get messed up if an error is detected in the middle of a message. It means that the next thing in the pipeline has to take (Either Error Result) instead of just Result, which is inelegant. I feel that, in general, if any stage in the pipeline has an error, the entire pipeline should be shut down. Figuring out exactly how to implement this is tricky; I decided to just use exceptions. Returning 'Done' as you suggest also makes sense, even if the user ends up losing information about why the pipeline was closed.

As you said, the semantics of (runGetPartial get) and the function returned by Partial are different. Because the latter is the common case, I feed the empty ByteString to the get just to have consistent semantics, so the 'push' function can always expect consistently behaving inputs, and not have to track if it should pass through an empty ByteString or disregard an empty ByteString. I feel that it makes the implementation simpler.

The one case where your implementation appears to break is when the Get doesn't consume any input and the first item is an empty ByteString. The expected output in this case would be whatever the Get produces, without consuming any input. I believe your implementation would return NeedInput. Your conduitGet seems to have the same problem - the expected output would be an infinite stream of whatever the Get produces without consuming any input, and I believe your implementation would return NeedsInput. I'll add tests that tests this.

Yes, you're completely right about returning the unconsumed input. My implementation of these functions are probably not right. Not only does the accumulator grow without bound, but each invocation of 'consume' calls BS.concat, which is O(n), and n grows, so overall it's O(n^2), which is awful. Get doesn't support readahead, so it doesn't have to store the input string itself - it only has to store the intermediate state of the Get. (For example, if my Get is (foldM (flip ($)) 0 $replicate 1000 $ \ x -> getWord8 >>= \ y -> return $ x + y) it doesn't have to save the whole 1000 character string)

This was my reasoning when designing the partial input part of cereal-conduit:
a) Conduit's connection operator knows which items it has passed to a sink/conduit, as well as those items it has not yet passed to the sink/conduit. When a sink completes (either successfully or not successfully), the partially-consumed data is going to be all the items the sink has ever consumed, plus some part of one item. In the case of early termination, early termination can only occur on an item boundary, so the "some part of one item" is always empty for this case, so I unconditionally return Nothing in sinkGet. This way, the Conduit infrastructure knows that the partially consumed data is just all the blocks it has ever passed to sinkGet.

b) However, the Data.Conduit infrastructure cannot make such an inference about conduitGet, because it does not know which part of which input corresponds to the beginning of the current Get. From the perspective of the Data.Conduit infrastructure, It appears that conduitGet has been happily accepting items all along, before finally receiving EOF. This is why I kept track of the ByteString that corresponds to the current Get invocation.

c) On the other hand, if the Get fails in sinkGet, I figure that the Data.Conduit infrastructure already knows that all the blocks it has handed to sinkGet are part of the partially consumed data. The only thing that the infrastructure doesn't know is where in the current item the failure occurred. That's why I only offer this information back up to Data.Conduit.

d) If the Get in conduitGet fails, then the previous two arguments both apply. The Data.Conduit infrastructure does not know where the current Get invocation started, and it doesn't know where in the current item the Get failed. Therefore, I have concatenate both the previously consumed data as well as part of the current item. I chose to throw an exception instead for the reasons in the first paragraph.

Please note that this is just my own reasoning - I don't know if Data.Conduit actually works this way. I haven't read the code for Data.Conduit. Perhaps I should include Michael Snoyman in this discussion?

---

As far as your particular application is concerned: I don't quite understand what you mean by "retranslate all messages from last successfully parsed one till current." UDP is allowed to re-order packets, so the last successfully parsed packet may be ahead of the failed packet. In this case, when you ask for a retransmission, the broken packet isn't in the range of the last successfully parsed packet till current.

It sounds like you're trying to add reliable delivery to conduits, which are a fundamentally stream-oriented structure. Reliable streams is something that TCP is really good at =P

Keep in mind that a service like you propose isn't scalable - If any receiver receives a corrupt packet, it will cause a retransmission (rebroadcast). As the number of receivers grows, the probability that a corrupted packet will be delivered grows as well. What if thousands of packets get sent, only to have a receiver say their packet #1 is corrupted? Will it retransmit all the thousands of packets? Or will there be a sliding window? If there's a sliding window, will it be kept per-receiver? This is starting to sound a lot like TCP.

If you have control over the sending machine and you're only interested in data consistency, you could use UDP's optional checksum. That would guarantee that the receiver OS shouldn't even deliver your Haskell program the packet if the checksum doesn't match. If that's not an option, it might make sense to make your conduit just drop the packet, like an OS would.

As for myself, I'm using cereal-conduit to parse text files from HTTP (using http-conduit) such as the public suffix list, at http://publicsuffix.org/list/. Eventually, I'll release this on Hackage.

**Comment by litherum** [comment]
I forgot to mention: I'm probably going to rewrite conduitGet using conduitState, modeled after http://hackage.haskell.org/packages/archive/conduit/0.4.0.1/doc/html/src/Data-Conduit-Binary.html#lines. That rewrite may make this discussion moot.

**Comment by schernichkin** [comment]
Thank you for the detailed response. I suppose first think we must clarify is how to handle empty bytestring. Because now it’s may be unobvious. I think it would be nice to have answers on following questions:
1.  Does Cereal really able to produce/consume an empty bytestring? How it will looks like? Will “runGetPartial get empty” return “Done” immediately, or it will return “Partial” which will return “Done” upon feeding empty string? Is it possible that “Get” result will depend on string length (for example “Get” will desterilize a kind of variable length string, where empty bytestring considered string terminator)? 
2.  Is stream containing two empty bytestrings equal to stream containing one empty bytestring? Will removing or inserting empty bytestrings change the meaning of whole stream?
3.  Suppose we have Get which can produce result from an empty bytestring. How we should handle input stream, probably containing several empty bytestream? Should we consume it all before producing result, should we produce result without any consumption (is how your implementation works, after you explanation I tend to think it’s correct and mine is wrong), should we consume exactly one chunk, or since inserting/deleting empty bytestring (possible) does not change meaning of the stream, we can do whatever we want and the only requirement is to produce result after final number of steps?

I also was rethinking my contracts and I found your idea about preserving partially consumed stream quite attractive. It makes consumption atomic – sink/conduit either consumes requited part of input and produce output or does not consume input at all and produce nothing. I’m going to add this behavior to my version of sink. Quadratic complexity is not a case, because we don’t have to concatenate string on each step, we may put in in list then revert it and use “ByteString.concat”, it will be  O(n). I also decided to replace “Either” with “Maybe” in sink, because the only useful information upstream is wherever I can deserialize data or not. Free-form error message Cereal will not help me much. I decided to call “Done” on conduit in case of deserialization error, I’m completely agree with your comment on stream synchronization. To give user an opportunity to handle errors I intending to provide two versions of “Get” conduits, something  like “sinkGetWithErrorHandler :: (String -> m ()) -> Get output -> ..” or even «((String, ByteString) -> Pipe ByteString output m <result>) -> ..» where first argument is monadic error handler. Shortened “sinkGet” will use predefined version (empty handler, or «Done (streamToMaybe unconsumeg) Nothing» in second case). I don’t see much benefits in using sinkState ect.. on my opinion it is just high level wrappers. I’m going to try all of this this week (before Monday) and will follow your suggestions, of course.

**Comment by litherum** [comment]
I believe the 'stream' is simply a way of slicing up a conceptual sequence of items, so I believe an empty item is conceptually the same as no item at all. As such,
sourceList [BS.pack [1], BS.empty, BS.pack [2]]
should behave the same as
sourceList [BS.empty, BS.empty, BS.pack [1], BS.empty, BS.pack [2]]
which should behave the same as
sourceList [BS.pack [1, 2]]
which should behave the same as
sourceList [BS.pack [1], BS.empty, BS.pack [2], BS.empty]

I also believe that a sink or conduit should not consume an item if it does not have to. In some cases, consuming an extra item may actually be incorrect if the stream ends at the end of an item.

(return () :: Get ()) is valid; that Get doesn't consume any input.

> runGetPartial (return ()) $ BS.singleton 65
> Done () "A"

This shows that (return()) does not consume any input, and the entire BS.singleton is returned as the leftover.

Combining these two beliefs leads to the conclusion that (conduitGet $ return ()) should result in an infinite stream of ()s as well as not consuming any input (0 items).

You're correct about ByteString.concat and O(n). I'll do this in v0.0.6.

I think your approach of having 2 conduits, one safe and one unsafe is a good idea. I'll probably do something like this in 0.0.6.

Currently, (I believe) there's no way to get partial data out of a conduit once you've used the $$ operator, so it's fairly unclear how to best handle this.

**Comment by litherum** [comment]
http://www.haskell.org/pipermail/haskell-cafe/2012-April/100715.html

**Comment by schernichkin** [comment]
Hi. Plz, take a look at my recent changes. I’ve created separated functions (mkConduidGet and mkSinkGet) in the Data.Conduit.Cereal.Internal, it allow to provide custom error handlers. I’ve also changed default sinkGet implementation now it uses MonadError just like conduitGet. I’m also considering MonadThrow but had not find out yet when it should be used exactly. If everything is ok I will make pull request.

**Comment by litherum** [comment]
I'm still emailing with Michael Snoyman about the best way to do partial data. As soon as it's clear, I'll be able to tell if this should be merged. After an initial pass, it looks pretty good.

**Comment by litherum** [comment]
Alright, your approach is good. Before you issue the pull request, you should rebase, and you should also format your your code to look like mine.

Once you make the pull request, I'll respond inline with a couple comments I have, but in general it looks good.
